### PR TITLE
fix include outbound ports iptable rules in tproxy mode

### PR DESCRIPTION
**Pull request by ramaraochavali**
As we are discussing https://github.com/istio/istio/issues/27565#issuecomment-765184403 we have an issue with `IncludeOutboundPorts` not working properly in `TPROXY` mode. 

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ X] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@lambdai see if this makes sense

**Review by rlenglet — CHANGES_REQUESTED** [review]
I don't think this makes sense. `ISTIO_OUTPUT` is always a `nat` chain.

**Comment by ramaraochavali** [comment]
@rlenglet I see. Thank you.  Based on the ip table output attached here https://github.com/istio/istio/issues/27565#issuecomment-764525722, would you be able to help what is the problem with `IncludeOutboundPorts` in `TPROXY` mode?

**Comment by lambdai** [comment]
The current tproxy chain in NAT table is below. see full table in https://github.com/istio/istio/issues/27565#issuecomment-764525722

I don't understand why neither ` -m owner --uid-owner 1337 -j RETURN` nor `-m owner --gid-owner 1337 -j RETURN` hit.
These two rules are supposed to bypass NAT rewrite(target ISTIO_REDIRECT) when Envoy connects to other containers.

@ramaraochavali confirmed that the Envoy is running under uid 1337 in tproxy?

@rlenglet Do you have any clue?

```
-A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -p tcp --dport 50010 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50075 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50070 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50090 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 9000 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8019 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8480 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8481 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8485 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 2181 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -d 127.1.2.3/32 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -j RETURN
```

**Comment by howardjohn** [comment]
is envoy running as uid 0 for tproxy for root?

**Comment by lambdai** [comment]
@howardjohn Theoretically the requirement is CAP_NET_ADMIN. Uid 0 is overkill.

**Comment by howardjohn** [comment]
I recall the injection template may set it to 0 though - worth checking. I
may remember wrong

**Comment by lambdai** [comment]
> I recall the injection template may set it to 0 though - worth checking. I may remember wrong
> […](#)

Then I guess the `PROXY_UID=1337, PROXY_GID=1337` alone doesn't imply envoy running uid 1337.

Maybe the injection deserves the change, not the iptables. 
CC @ramaraochavali

**Comment by ramaraochavali** [comment]
@lambdai Thank you so much for looking in to this. Yeah, I also figured out that it is not returning properly but could not think uid is problem - I was tripped off by the iptable output. Looked at the injection template and pod.yaml - Yes, in TPROXY mode it is running as 0 as @howardjohn says. 

I reverted the iptable change and fixed the template. 

@lambdai @howardjohn @rlenglet PTAL - Thanks for helping out on this.

File: manifests/charts/istio-control/istio-discovery/files/injection-template.yaml [hunk 1]
```diff
@@ -343,8 +343,14 @@ spec:
...
       fsGroup: 1337
       {{ if or (eq (annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode) `TPROXY`) (eq (annotation .ObjectMeta `sidecar.istio.io/capNetBindService` .Values.global.proxy.capNetBindService) `true`) -}}
+      {{ if eq (annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode) `TPROXY` -}}
+      runAsNonRoot: false
+      runAsUser: 1337
+      {{- end }}
+      {{ if eq (annotation .ObjectMeta `sidecar.istio.io/capNetBindService` .Values.global.proxy.capNetBindService) `true` -}}
       runAsNonRoot: false
       runAsUser: 0
+      {{- end }}
```
Comment (lambdai) [inline manifests/charts/istio-control/istio-discovery/files/injection-template.yaml:353]: These 2 `if` are not exclusive. IT means runAsNonRoot and runAsUser could appear twice.

Forgive my lack of knowledge on template language: what is the corresponding `if` of the `else` on line 354?

File: pkg/kube/inject/testdata/inject/hello-tproxy.yaml.injected [hunk 2]
```diff
@@ -137,7 +137,7 @@ spec:
           readOnlyRootFilesystem: true
           runAsGroup: 1337
```
Comment (lambdai) [inline pkg/kube/inject/testdata/inject/hello-tproxy.yaml.injected:138]: See here: gid 1337 should have the same effect as uid 1337.

File: manifests/charts/istiod-remote/files/injection-template.yaml [hunk 3]
```diff
@@ -343,8 +343,14 @@ spec:
...
       fsGroup: 1337
       {{ if or (eq (annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode) `TPROXY`) (eq (annotation .ObjectMeta `sidecar.istio.io/capNetBindService` .Values.global.proxy.capNetBindService) `true`) -}}
+      {{ if eq (annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode) `TPROXY` -}}
+      runAsNonRoot: false
+      runAsUser: 1337
+      {{- end }}
+      {{ if eq (annotation .ObjectMeta `sidecar.istio.io/capNetBindService` .Values.global.proxy.capNetBindService) `true` -}}
       runAsNonRoot: false
       runAsUser: 0
+      {{- end }}
```
Comment (lambdai) [inline manifests/charts/istiod-remote/files/injection-template.yaml:353]: ditto

**Review by lambdai — CHANGES_REQUESTED** [review]
Great! We are on the same page. The uid/gid is the culprit.

I am looking at the hello-tproxy.yaml. The golden test has the `runAsGroup: 1337` that is supposed to match iptables rules.

I am wondering if `runAsGroup:1337` doesn't work as we expected. Maybe `securityContext` as whole, or docker image managing linux [user namespaces](https://man7.org/linux/man-pages/man7/user_namespaces.7.html) was broken.

Anyway, our ultimate goal is to run Envoy 
1. under either uid 1337 or gid 1337, or both.
2. Envoy process must have CAP_NET_ADMIN capability if TPROXY is enabled.

File: pkg/kube/inject/testdata/inject/hello-tproxy.yaml.injected (hunk 2 above)
Comment (lambdai) [reply pkg/kube/inject/testdata/inject/hello-tproxy.yaml.injected:138]: This is not the common output of tproxy setting but a special arg in the test case

File: manifests/charts/istio-control/istio-discovery/files/injection-template.yaml (hunk 1 above)
Comment (lambdai) [reply manifests/charts/istio-control/istio-discovery/files/injection-template.yaml:353]: Ok, the else matches the `if` w/o `end`

**Comment by ramaraochavali** [comment]
@lambdai So we patch the security context in the operator with the following
```
securityContext:
  allowPrivilegeEscalation: false
  capabilities:
    drop:
    - ALL
  privileged: false
  readOnlyRootFilesystem: false
  runAsUser: 1337
```
And the SecurityContext for `TPROXY` pod is
```
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        add:
        - NET_ADMIN
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsNonRoot: false
      runAsUser: 0
```  
and for `REDIRECT` is 

```
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsNonRoot: true
      runAsUser: 1337
```

So I guess the operator patching is ignoring `runAsGroup` as it is not there in the patch?

So I am wondering the following rule is helping in `REDIRECT` mode 
```
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN 
```
But since runAsUser is 0 for TPROXY - it is failing.

So this PR is setting UID to 1337 which works for us.

WDYT?

File: manifests/charts/istio-control/istio-discovery/files/injection-template.yaml (hunk 1 above)
Comment (ramaraochavali) [reply manifests/charts/istio-control/istio-discovery/files/injection-template.yaml:353]: Ah. Will fix.

**Comment by rlenglet** [comment]
When did `runAsGroup` ever get removed from the `securityContext`? 🤔

**Review by rlenglet — CHANGES_REQUESTED** [review]
Let's first understand why `runAsGroup` is not set in the `securityContext` of the container. It's specified in the injection template, but disappeared in the container's context. I don't understand how that's possible.

**Comment by ramaraochavali** [comment]
@rlenglet Agreed.  I am still looking in to why `runAsGroup` is not coming even when the injection template has it. One possible explanation is `runAsGroup` on container can be disabled using feature gate https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/. I am checking with our kubernetes team if that is the case in our environment.

Do you see any issues setting user to `1337` for `TPROXY`?

**Comment by howardjohn** [comment]
If UID != GID there are some edge cases that may not work btw. I *think* TestTraffic/selfcall will fail

**Comment by rlenglet** [comment]
> Do you see any issues setting user to 1337 for TPROXY?

Yes. IIRC, that's because Docker doesn't set ambient capabilities. So when it switches the container's UID from 0 to 1337, it loses all capabilities. Adding capabilities only work if UID is 0.

**Comment by lambdai** [comment]
> If UID != GID there are some edge cases that may not work btw. I _think_ TestTraffic/selfcall will fail

That's a good point. The goals of (uid != 1337, gid=1337) are 
1. Don't intercept uid=0 since we assume this uid is shared among Envoy and other processes.
2. Capture on gid=1337 which is Envoy(or whichever don't want to be captured)

So the Envoy runAsUid and captured uid must be independent, with the default case that 
runAsUid = 1337, captured uid = {1337}. Note that captured uid is a collection.

**Comment by lambdai** [comment]
> Yes. IIRC, that's because Docker doesn't set ambient capabilities. So when it switches the container's UID from 0 to 1337, it loses all capabilities. Adding capabilities only works if the UID is 0.

Is this the capability uniquely lost in docker-shim CRI?  Other CRI may support

**Comment by ramaraochavali** [comment]
Ok. Another finding to test whether gid is the actual issue, I passed `-g 0,1337` to iptables to let it skip gid `0` also. But the problem exists - It makes me wonder it is not gid but some thing else is there

Here is the proxy id `uid=0(root) gid=0(root) groups=0(root),1337(istio-proxy)`

and the ip table rules - now it added a rule to skip if gid is 0 - but still it did not skip

```
Environment:
------------
ENVOY_PORT=
INBOUND_CAPTURE_PORT=
ISTIO_INBOUND_INTERCEPTION_MODE=
ISTIO_INBOUND_TPROXY_MARK=
ISTIO_INBOUND_TPROXY_ROUTE_TABLE=
ISTIO_INBOUND_PORTS=
ISTIO_OUTBOUND_PORTS=
ISTIO_LOCAL_EXCLUDE_PORTS=
ISTIO_SERVICE_CIDR=
ISTIO_SERVICE_EXCLUDE_CIDR=

Variables:
----------
PROXY_PORT=15001
PROXY_INBOUND_CAPTURE_PORT=15006
PROXY_TUNNEL_PORT=15008
PROXY_UID=1337
PROXY_GID=0,1337
INBOUND_INTERCEPTION_MODE=TPROXY
INBOUND_TPROXY_MARK=1337
INBOUND_TPROXY_ROUTE_TABLE=133
INBOUND_PORTS_INCLUDE=*
INBOUND_PORTS_EXCLUDE=15090,15021,8181,15020
OUTBOUND_IP_RANGES_INCLUDE=127.1.2.3/32
OUTBOUND_IP_RANGES_EXCLUDE=
OUTBOUND_PORTS_INCLUDE=50010,50020,50075,50070,50090,9000,8019,8480,8481,8485,2181,8030,8031,8032,8033,8088,8040,8041,8042,10020,19888,33911,8020,8082
OUTBOUND_PORTS_EXCLUDE=
KUBEVIRT_INTERFACES=
ENABLE_INBOUND_IPV6=false

ip -f inet rule add fwmark 1337 lookup 133
ip -f inet route add local default dev lo table 133
Writing following contents to rules file:  /tmp/iptables-rules-1611637611221538655.txt201772342
* nat
-N ISTIO_INBOUND
-N ISTIO_REDIRECT
-N ISTIO_IN_REDIRECT
-N ISTIO_OUTPUT
-A ISTIO_INBOUND -p tcp --dport 15008 -j RETURN
-A ISTIO_REDIRECT -p tcp --dport 53 -j REDIRECT --to-ports 15053
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 0 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 0 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -p tcp --dport 50010 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50075 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50070 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50090 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 9000 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8019 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8480 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8481 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8485 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 2181 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8030 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8031 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8032 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8033 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8088 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8040 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8041 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8042 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 10020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 19888 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 33911 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8082 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -d 127.1.2.3/32 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -j RETURN
-A OUTPUT -p udp --dport 53 -m owner --uid-owner 1337 -j RETURN
-A OUTPUT -p udp --dport 53 -m owner --gid-owner 0 -j RETURN
-A OUTPUT -p udp --dport 53 -m owner --gid-owner 1337 -j RETURN
-A OUTPUT -p udp --dport 53 -j DNAT --to-destination 127.0.0.1:15053
-A POSTROUTING -p udp --dport 15053 -j SNAT --to-source 127.0.0.1
COMMIT
* mangle
-N ISTIO_DIVERT
-N ISTIO_TPROXY
-N ISTIO_INBOUND
-A ISTIO_DIVERT -j MARK --set-mark 1337
-A ISTIO_DIVERT -j ACCEPT
-A ISTIO_TPROXY ! -d 127.0.0.1/32 -p tcp -j TPROXY --tproxy-mark 1337/0xffffffff --on-port 15006
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A ISTIO_INBOUND -p tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15021 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 8181 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -m conntrack --ctstate RELATED,ESTABLISHED -j ISTIO_DIVERT
-A ISTIO_INBOUND -p tcp -j ISTIO_TPROXY
-A PREROUTING -p tcp -m mark --mark 1337 -j CONNMARK --save-mark
-A OUTPUT -p tcp -m connmark --mark 1337 -j CONNMARK --restore-mark
-I ISTIO_INBOUND 1 -p tcp -m mark --mark 1337 -j RETURN
COMMIT

iptables-restore --noflush /tmp/iptables-rules-1611637611221538655.txt201772342
Writing following contents to rules file:  /tmp/ip6tables-rules-1611637611969044794.txt772360989

ip6tables-restore --noflush /tmp/ip6tables-rules-1611637611969044794.txt772360989
iptables-save 
# Generated by iptables-save v1.4.21 on Tue Jan 26 05:06:51 2021
*mangle
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:ISTIO_DIVERT - [0:0]
:ISTIO_INBOUND - [0:0]
:ISTIO_TPROXY - [0:0]
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A PREROUTING -p tcp -m mark --mark 0x539 -j CONNMARK --save-mark --nfmask 0xffffffff --ctmask 0xffffffff
-A OUTPUT -p tcp -m connmark --mark 0x539 -j CONNMARK --restore-mark --nfmask 0xffffffff --ctmask 0xffffffff
-A ISTIO_DIVERT -j MARK --set-xmark 0x539/0xffffffff
-A ISTIO_DIVERT -j ACCEPT
-A ISTIO_INBOUND -p tcp -m mark --mark 0x539 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15021 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 8181 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -m conntrack --ctstate RELATED,ESTABLISHED -j ISTIO_DIVERT
-A ISTIO_INBOUND -p tcp -j ISTIO_TPROXY
-A ISTIO_TPROXY ! -d 127.0.0.1/32 -p tcp -j TPROXY --on-port 15006 --on-ip 0.0.0.0 --tproxy-mark 0x539/0xffffffff
COMMIT
# Completed on Tue Jan 26 05:06:51 2021
# Generated by iptables-save v1.4.21 on Tue Jan 26 05:06:51 2021
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:ISTIO_INBOUND - [0:0]
:ISTIO_IN_REDIRECT - [0:0]
:ISTIO_OUTPUT - [0:0]
:ISTIO_REDIRECT - [0:0]
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A OUTPUT -p udp -m udp --dport 53 -m owner --uid-owner 1337 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -m owner --gid-owner 0 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -m owner --gid-owner 1337 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -j DNAT --to-destination 127.0.0.1:15053
-A POSTROUTING -p udp -m udp --dport 15053 -j SNAT --to-source 127.0.0.1
-A ISTIO_INBOUND -p tcp -m tcp --dport 15008 -j RETURN
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A ISTIO_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 0 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 0 -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50010 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50075 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50070 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50090 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 9000 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8019 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8480 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8481 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8485 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 2181 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8030 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8031 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8032 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8033 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8088 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8040 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8041 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8042 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 10020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 19888 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 33911 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8082 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -d 127.1.2.3/32 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -j RETURN
-A ISTIO_REDIRECT -p tcp -m tcp --dport 53 -j REDIRECT --to-ports 15053
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
COMMIT
# Completed on Tue Jan 26 05:06:51 2021
```

**Comment by ramaraochavali** [comment]
So what do we do with this PR? I am still figuring out why `runAsGroup` as not being added in our environment even though injection template has it. Do we want to make `uid` and `gid` for the reasons @howardjohn mentioned or just rely on `gid` to skip?

**Comment by rlenglet** [comment]
> Is this the capability uniquely lost in docker-shim CRI? Other CRI may support

No, I think none of them sets ambient capabilities. We can't rely on that.

**Comment by rlenglet** [comment]
@ramaraochavali the problem seems to be specific to your environment, so I would like to first understand what's going on there before we commit anything. Thanks!

**Comment by ramaraochavali** [comment]
@rlenglet Sure. That makes sense. I will update what I find here.

**Comment by ramaraochavali** [comment]
I will close this PR for now and once we find what is causing `runAsGroup` not showing up and if it needs a fix, will open a separate PR.

**Comment by ramaraochavali** [comment]
@rlenglet @lambdai Update. The "runAsGroup" not setting per container seems to be problem with EKS. Our K8s team is a following up on that.

Another question - Do we really need to be "root" (runAsNonRoot: false)  for "TPROXY" or just "NET_ADMIN" is sufficient? I think since we are running as root, the gid is automatically set to 0 and since runAsGroup is not working as expected in our env, we are stuck with gid 0. If root is really not needed, removing that also would solve our case.

**Comment by rlenglet** [comment]
root is needed. Again, adding capabilities doesn't work in container runtimes because they don't set ambient capabilities, so capabilities are lost when changing the uid. So uid 0 is required in order to preserve `CAP_NET_ADMIN`.

**Comment by rlenglet** [comment]
@ramaraochavali Do you have any reference about `runAsGroup` not working on EKS?

**Comment by ramaraochavali** [comment]
@rlenglet No there is no reference any where. We could find by experiment and we did not get any response yet.

**Comment by ramaraochavali** [comment]
@rlenglet Just closing loop on this, it turned out that there is a bug in one of the mutating webhooks that is running in our cluster is pinned to older version of client-go that does have "RunAsGroup" in the SecurityContext so was getting dropped during mutation. The webhook owner is fixing it. So It is not an AWS issue either.

**Comment by rlenglet** [comment]
@ramaraochavali awesome, thanks for looking deeper into this!

**Pull request closed without merging**